test/studies/ml/lib/Chai.chpl

```diff
@@ -0,0 +1,828 @@
+
+            weightsGrad += dL_dW;
+            biasesGrad += dL_dB;
+
+            return dL_dIn.reshape(input.domain);
+        }
+
+        proc backwardBatch(deltas: [] Tensor(1), inputs: [] ?tensorType2) : [] tensorType2 where isSubtype(tensorType2, Tensor) {
+            const batchSize = deltas.size;
+            var newDeltas: [0..#batchSize] tensorType2;
+            newDeltas.reshapeDomain(inputs[0].domain);
+
+            var weightsGrad = this.weightsGrad;
+            var biasesGrad = this.biasesGrad;
+            forall (delta,input,idx) in zip(deltas,inputs,newDeltas.domain) with (ref this, ref deltas, ref inputs, ref newDeltas, + reduce weightsGrad, + reduce biasesGrad) {
+                // Coppied from above
+                const flattened = input.flatten();
+                const Z = (weights * flattened) + biases;
+                const (exp,expSum,softmax) = tn.softmaxParts(Z);
+                const dL_dOut = delta;
+                var nonZeroIdx: int = -1;
+                for i in delta.data.domain do
+                    if delta[i] != 0.0 { nonZeroIdx = i; break; }
+                if nonZeroIdx == -1 then tn.err("Softmax backward: delta is zero vector");
+                const i = nonZeroIdx;
+                var dOut_dZ: Tensor(1) = (- exp[i]) * (exp / (expSum ** 2.0));
+                dOut_dZ[i] = exp[i] * (expSum - exp[i]) / (expSum ** 2.0);
+                const dZ_dW: Tensor(1) = flattened;
+                const dZ_dB: real = 1;
+                const dZ_dIn: Tensor(2) = weights;
+                const dL_dZ: Tensor(1) = dL_dOut[i] * dOut_dZ;
+                const dL_dW: Tensor(2) = dL_dZ * dZ_dW.transpose();
+                const dL_dB: Tensor(1) = dL_dZ * dZ_dB;
+                const dL_dIn: Tensor(1) = dZ_dIn.transpose() * dL_dZ;
+                weightsGrad += dL_dW;
+                biasesGrad += dL_dB;
+                newDeltas[idx] = dL_dIn.reshape(input.domain);;
+            }
+            this.weightsGrad.data = weightsGrad.data;
+            this.biasesGrad.data = biasesGrad.data;
+            return newDeltas;
+        }
+
+        proc optimize(mag: real(64)) {
+            weights.data -= mag * weightsGrad.data;
+            biases.data -= mag * biasesGrad.data;
+        }
+        proc resetGradients() {
+            weightsGrad.data = 0.0;
+            biasesGrad.data = 0.0;
+        }
+
+        proc write(fw: IO.fileWriter) throws {
+            weights.write(fw);
+            biases.write(fw);
+        }
+        proc read(fr: IO.fileReader) throws {
+            weights.read(fr);
+            biases.read(fr);
+            uninitialized = false;
+        }
+
+        proc signature(): string {
+            return "SoftMax(" + outputSize:string + ")";
+        }
+
+    }
+
+
+    proc forwardPropHelp(ref layers, param n: int, x: Tensor(?)) {
+        if n == layers.size then return x;
+
+        const xNext = layers[n].forwardProp(x);
+        return forwardPropHelp(layers, n+1, xNext);
```

Comment:
The fact that we had to use recursion here instead of a loop was a bit unusual. Would you mind adding a comment here that explains why a simple loop wasn't working with `xNext` having a potentially different type each iteration?